boutdata/collect.py

```diff
@@ -307,392 +236,776 @@ def getDataFile(i):
-            print("WARNING: More files than expected (" + str(npe) + ")")
-        elif npe > nfiles:
-            print("WARNING: Some files missing. Expected " + str(npe))
+        myg = 0
+        print("MYG not found, setting to {}".format(myg))
 
     if xguards:
-        nx = nxpe * mxsub + 2*mxg
+        nx = f["nx"]
     else:
-        nx = nxpe * mxsub
-
+        nx = f["nx"] - 2 * mxg
     if yguards:
-        ny = mysub * nype + 2*myg
+        ny = f["ny"] + 2 * myg
         if yguards == "include_upper" and f["jyseps2_1"] != f["jyseps1_2"]:
             # Simulation has a second (upper) target, with a second set of y-boundary
             # points
-            ny = ny + 2*myg
-            ny_inner = f["ny_inner"]
-            yproc_upper_target = ny_inner // mysub - 1
-            if f["ny_inner"] % mysub != 0:
-                raise ValueError("Trying to keep upper boundary cells but "
-                                 "mysub={} does not divide ny_inner={}"
-                                 .format(mysub, ny_inner))
-        else:
-            yproc_upper_target = None
+            ny = ny + 2 * myg
     else:
-        ny = mysub * nype
+        ny = f["ny"]
+    nz = f["MZ"]
+    t_array = f.read("t_array")
+    if t_array is None:
+        nt = 1
+        t_array = np.zeros(1)
+    else:
+        try:
+            nt = len(t_array)
+        except TypeError:
+            # t_array is not an array here, which probably means it was a
+            # one-element array and has been read as a scalar.
+            nt = 1
 
     xind = _convert_to_nice_slice(xind, nx, "xind")
     yind = _convert_to_nice_slice(yind, ny, "yind")
     zind = _convert_to_nice_slice(zind, nz, "zind")
     tind = _convert_to_nice_slice(tind, nt, "tind")
 
-    xsize = xind.stop - xind.start
-    ysize = yind.stop - yind.start
-    zsize = int(np.ceil(float(zind.stop - zind.start)/zind.step))
-    tsize = int(np.ceil(float(tind.stop - tind.start)/tind.step))
+    if not xguards:
+        xind = slice(xind.start + mxg, xind.stop + mxg, xind.step)
+    if not yguards:
+        yind = slice(yind.start + myg, yind.stop + myg, yind.step)
+
+    if dimensions == ():
+        ranges = []
+    elif dimensions == ("t",):
+        ranges = [tind]
+    elif dimensions == ("x", "y"):
+        # Field2D
+        ranges = [xind, yind]
+    elif dimensions == ("x", "z"):
+        # FieldPerp
+        ranges = [xind, zind]
+    elif dimensions == ("t", "x", "y"):
+        # evolving Field2D
+        ranges = [tind, xind, yind]
+    elif dimensions == ("t", "x", "z"):
+        # evolving FieldPerp
+        ranges = [tind, xind, zind]
+    elif dimensions == ("x", "y", "z"):
+        # Field3D
+        ranges = [xind, yind, zind]
+    elif dimensions == ("t", "x", "y", "z"):
+        # evolving Field3D
+        ranges = [tind, xind, yind, zind]
+    else:
+        # Not a Field, so do not support slicing.
+        # For example, may be a string.
+        ranges = None
```

Comment:
It could be, but the function would only be used in one place - slicing has to be applied differently when collecting from multiple files (where the 'step' is applied after collecting') from collecting from a single file (all slicing can be done while reading). The function body is only 86 lines long anyway, so I don't think it's worth splitting to a separate function just to split the function up.

Reply:
Here's a suggestion:

```suggestion
    dim_ranges = {"t": tind, "x": xind, "y": yind, "z": zind}
    ranges = [dim_ranges.get(dim, None) for dim in dimensions]

    if None in ranges:
        # Not a Field, so do not support slicing
        # For example, may be a string
        ranges = None
```

I think this covers all the cases, plus some theoretical cases that could still support slicing like `("t", "x")`  :smile:

Reply:
Neat! I think we could actually drop the second condition - if a variable has some time/space dimensions plus some other dimensions, it does make sense to slice the time/space, and a 'range' entry of `None` for the others will do no slicing.
```suggestion
    dim_ranges = {"t": tind, "x": xind, "y": yind, "z": zind}
    ranges = [dim_ranges.get(dim, None) for dim in dimensions]
```